// bee-proto/dist/proto/code_interpreter/v1/code_interpreter_service_connect.cjs
/**
 * Copyright 2024 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var code_interpreter_service_connect_exports = {};
__export(code_interpreter_service_connect_exports, {
  CodeInterpreterService: () => CodeInterpreterService
});
module.exports = __toCommonJS(code_interpreter_service_connect_exports);
var import_code_interpreter_service_pb = require("./code_interpreter_service_pb.js");
var import_protobuf = require("@bufbuild/protobuf");
const CodeInterpreterService = {
  typeName: "code_interpreter.v1.CodeInterpreterService",
  methods: {
    /**
     * @generated from rpc code_interpreter.v1.CodeInterpreterService.Execute
     */
    execute: {
      name: "Execute",
      I: import_code_interpreter_service_pb.ExecuteRequest,
      O: import_code_interpreter_service_pb.ExecuteResponse,
      kind: import_protobuf.MethodKind.Unary
    },
    /**
     * @generated from rpc code_interpreter.v1.CodeInterpreterService.ParseCustomTool
     */
    parseCustomTool: {
      name: "ParseCustomTool",
      I: import_code_interpreter_service_pb.ParseCustomToolRequest,
      O: import_code_interpreter_service_pb.ParseCustomToolResponse,
      kind: import_protobuf.MethodKind.Unary
    },
    /**
     * @generated from rpc code_interpreter.v1.CodeInterpreterService.ExecuteCustomTool
     */
    executeCustomTool: {
      name: "ExecuteCustomTool",
      I: import_code_interpreter_service_pb.ExecuteCustomToolRequest,
      O: import_code_interpreter_service_pb.ExecuteCustomToolResponse,
      kind: import_protobuf.MethodKind.Unary
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CodeInterpreterService
});
