// bee-agent-framework/dist/adapters/ollama/llm.cjs
'use strict';

var llm_cjs = require('../../llms/llm.cjs');
var emitter_cjs = require('../../emitter/emitter.cjs');
var base_cjs = require('../../llms/base.cjs');
var ollama = require('ollama');
var decoratorCache_cjs = require('../../cache/decoratorCache.cjs');
var number_cjs = require('../../internals/helpers/number.cjs');
var utils_cjs = require('../../serializer/utils.cjs');
var promise_cjs = require('../../internals/helpers/promise.cjs');
var object_cjs = require('../../internals/helpers/object.cjs');
var shared_cjs = require('./shared.cjs');
var env_cjs = require('../../internals/env.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
function _ts_decorate(decorators, target, key, desc) {
  var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
  else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
__name(_ts_decorate, "_ts_decorate");
function _ts_metadata(k, v) {
  if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
}
__name(_ts_metadata, "_ts_metadata");
class OllamaLLMOutput extends base_cjs.BaseLLMOutput {
  static {
    __name(this, "OllamaLLMOutput");
  }
  results;
  constructor(result) {
    super();
    this.results = [
      result
    ];
  }
  static {
    this.register();
  }
  getTextContent() {
    return this.finalResult.response;
  }
  get finalResult() {
    if (this.results.length === 0) {
      throw new base_cjs.LLMOutputError("No chunks to get final result from!");
    }
    return object_cjs.customMerge(this.results, {
      response: /* @__PURE__ */ __name((value = "", oldValue = "") => oldValue + value, "response"),
      total_duration: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "total_duration"),
      load_duration: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "load_duration"),
      model: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "model"),
      done: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "done"),
      done_reason: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "done_reason"),
      created_at: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "created_at"),
      eval_duration: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "eval_duration"),
      prompt_eval_duration: /* @__PURE__ */ __name((value, oldValue) => value ?? oldValue, "prompt_eval_duration"),
      prompt_eval_count: number_cjs.safeSum,
      eval_count: number_cjs.safeSum,
      context: /* @__PURE__ */ __name((value, oldValue) => [
        ...value || [],
        ...oldValue || []
      ], "context")
    });
  }
  merge(other) {
    decoratorCache_cjs.Cache.getInstance(this, "finalResult").clear();
    this.results.push(...other.results);
  }
  createSnapshot() {
    return {
      results: utils_cjs.shallowCopy(this.results)
    };
  }
  loadSnapshot(snapshot) {
    Object.assign(this, snapshot);
  }
  toString() {
    return this.getTextContent();
  }
}
_ts_decorate([
  decoratorCache_cjs.Cache(),
  _ts_metadata("design:type", typeof Readonly === "undefined" ? Object : Readonly),
  _ts_metadata("design:paramtypes", [])
], OllamaLLMOutput.prototype, "finalResult", null);
class OllamaLLM extends llm_cjs.LLM {
  static {
    __name(this, "OllamaLLM");
  }
  emitter = emitter_cjs.Emitter.root.child({
    namespace: [
      "ollama",
      "llm"
    ],
    creator: this
  });
  client;
  parameters;
  static {
    this.register();
    shared_cjs.registerClient();
  }
  constructor({ client, modelId, parameters, executionOptions = {}, cache }) {
    super(modelId, executionOptions, cache);
    this.client = client ?? new ollama.Ollama({
      host: env_cjs.getEnv("OLLAMA_HOST")
    });
    this.parameters = parameters ?? {};
  }
  async _generate(input, options, run) {
    const response = await promise_cjs.signalRace(async () => this.client.generate({
      ...await this.prepareParameters(input, options),
      stream: false
    }), run.signal, () => this.client.abort());
    return new OllamaLLMOutput(response);
  }
  async *_stream(input, options, run) {
    for await (const chunk of await this.client.generate({
      ...await this.prepareParameters(input, options),
      stream: true
    })) {
      if (run.signal.aborted) {
        break;
      }
      yield new OllamaLLMOutput(chunk);
    }
    run.signal.throwIfAborted();
  }
  async version() {
    const config = object_cjs.getPropStrict(this.client, "config");
    return shared_cjs.retrieveVersion(config.host, config.fetch);
  }
  async meta() {
    const model = await this.client.show({
      model: this.modelId
    });
    return shared_cjs.extractModelMeta(model);
  }
  async embed(input, options = {}) {
    const response = await this.client.embed({
      model: this.modelId,
      input,
      options: options?.options,
      truncate: options?.truncate
    });
    return {
      embeddings: response.embeddings
    };
  }
  async tokenize(input) {
    return {
      tokensCount: Math.ceil(input.length / 4)
    };
  }
  async prepareParameters(input, overrides) {
    return {
      model: this.modelId,
      prompt: input,
      raw: true,
      options: this.parameters,
      format: shared_cjs.retrieveFormat(await this.version(), overrides?.guided)
    };
  }
  createSnapshot() {
    return {
      ...super.createSnapshot(),
      modelId: this.modelId,
      executionOptions: utils_cjs.shallowCopy(this.executionOptions),
      parameters: utils_cjs.shallowCopy(this.parameters),
      client: this.client
    };
  }
}
_ts_decorate([
  decoratorCache_cjs.Cache(),
  _ts_metadata("design:type", Function),
  _ts_metadata("design:paramtypes", []),
  _ts_metadata("design:returntype", Promise)
], OllamaLLM.prototype, "version", null);

exports.OllamaLLM = OllamaLLM;
exports.OllamaLLMOutput = OllamaLLMOutput;
