// bee-agent-framework/dist/adapters/bam/chatPreset.js
import { LLMChatTemplates } from '../shared/llmChatTemplates.js';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const BAMChatLLMPreset = {
  "meta-llama/llama-3-3-70b-instruct": /* @__PURE__ */ __name(() => {
    const { template, parameters, messagesToPrompt } = LLMChatTemplates.get("llama3.3");
    return {
      base: {
        parameters: {
          decoding_method: "greedy",
          include_stop_sequence: false,
          max_new_tokens: 2048,
          repetition_penalty: 1,
          stop_sequences: [
            ...parameters.stop_sequence
          ]
        }
      },
      chat: {
        messagesToPrompt: messagesToPrompt(template)
      }
    };
  }, "meta-llama/llama-3-3-70b-instruct"),
  "meta-llama/llama-3-1-70b-instruct": /* @__PURE__ */ __name(() => {
    const { template, parameters, messagesToPrompt } = LLMChatTemplates.get("llama3.1");
    return {
      base: {
        parameters: {
          decoding_method: "greedy",
          include_stop_sequence: false,
          max_new_tokens: 2048,
          repetition_penalty: 1,
          stop_sequences: [
            ...parameters.stop_sequence
          ]
        }
      },
      chat: {
        messagesToPrompt: messagesToPrompt(template)
      }
    };
  }, "meta-llama/llama-3-1-70b-instruct"),
  "meta-llama/llama-3-8b-instruct"() {
    return BAMChatLLMPreset["meta-llama/llama-3-70b-instruct"]();
  },
  "meta-llama/llama-3-70b-instruct": /* @__PURE__ */ __name(() => {
    const { template, parameters, messagesToPrompt } = LLMChatTemplates.get("llama3");
    return {
      base: {
        parameters: {
          decoding_method: "greedy",
          max_new_tokens: 1500,
          include_stop_sequence: false,
          stop_sequences: [
            ...parameters.stop_sequence
          ]
        }
      },
      chat: {
        messagesToPrompt: messagesToPrompt(template)
      }
    };
  }, "meta-llama/llama-3-70b-instruct")
};

export { BAMChatLLMPreset };
