// bee-agent-framework/dist/internals/serializable.js
import { Serializer } from '../serializer/serializer.js';
import * as R from 'remeda';
import { extractClassName } from '../serializer/utils.js';
import { SerializerError } from '../serializer/error.js';
import { Cache } from '../cache/decoratorCache.js';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class Serializable {
  static {
    __name(this, "Serializable");
  }
  constructor() {
    Object.getPrototypeOf(this).constructor.register();
    Cache.init(this);
  }
  static register(aliases) {
    Serializer.registerSerializable(this, void 0, aliases);
  }
  clone() {
    const snapshot = this.createSnapshot();
    const target = Object.create(this.constructor.prototype);
    target.loadSnapshot(snapshot);
    return target;
  }
  serialize() {
    const snapshot = this.createSnapshot();
    return Serializer.serialize({
      target: extractClassName(this),
      snapshot
    });
  }
  deserialize(value, options) {
    const { __root } = Serializer.deserializeWithMeta(value, options?.extraClasses);
    if (!__root.target) {
      console.warn(`Serializable class must be serialized via "serialize" method and not via Serializer class. This may lead to incorrect deserialization.`);
      return __root;
    }
    const current = extractClassName(this);
    if (current !== __root.target) {
      throw new SerializerError(`Snapshot has been created for class '${__root.target}' but you want to use it for class '${current}'.`);
    }
    return __root.snapshot;
  }
  static fromSnapshot(state) {
    const target = Object.create(this.prototype);
    target.loadSnapshot(state);
    Cache.init(target);
    return target;
  }
  static fromSerialized(serialized, options = {}) {
    const target = Object.create(this.prototype);
    const state = target.deserialize(serialized, options);
    const load = target.loadSnapshot(state);
    Cache.init(target);
    return R.isPromise(load) ? load.then(() => target) : target;
  }
}

export { Serializable };
