// bee-agent-framework/dist/instrumentation/create-telemetry-middleware.cjs
'use strict';

var getSerializedObjectSafe_cjs = require('./helpers/get-serialized-object-safe.cjs');
var createSpan_cjs = require('./helpers/create-span.cjs');
var idNameManager_cjs = require('./helpers/id-name-manager.cjs');
var getErrorSafe_cjs = require('./helpers/get-error-safe.cjs');
var remeda = require('remeda');
var errors_cjs = require('../errors.cjs');
var version_cjs = require('../version.cjs');
var message_cjs = require('../llms/primitives/message.cjs');
var tracer_cjs = require('./tracer.cjs');
var traceSerializer_cjs = require('./helpers/trace-serializer.cjs');
var config_cjs = require('./config.cjs');
var utils_cjs = require('../emitter/utils.cjs');
var logger_cjs = require('./logger.cjs');
var base_cjs = require('../agents/base.cjs');
var utils_cjs$1 = require('./helpers/utils.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
function createTelemetryMiddleware() {
  return (context) => {
    if (!context.emitter?.trace?.id) {
      throw new errors_cjs.FrameworkError(`Fatal error. Missing traceId`, [], {
        context
      });
    }
    const traceId = context.emitter?.trace?.id;
    if (tracer_cjs.activeTracesMap.has(traceId)) {
      return;
    }
    tracer_cjs.activeTracesMap.set(traceId, context.instance.constructor.name);
    logger_cjs.instrumentationLogger.debug({
      source: context.instance.constructor.name,
      traceId
    }, "createTelemetryMiddleware");
    const { emitter, runParams, instance } = context;
    const basePath = utils_cjs.createFullPath(emitter.namespace, "");
    let prompt = null;
    if (instance instanceof base_cjs.BaseAgent) {
      prompt = runParams[0].prompt;
    }
    const spansMap = /* @__PURE__ */ new Map();
    const parentIdsMap = /* @__PURE__ */ new Map();
    const spansToDeleteMap = /* @__PURE__ */ new Map();
    let generatedMessage = void 0;
    let history = void 0;
    const groupIterations = [];
    const idNameManager = new idNameManager_cjs.IdNameManager();
    const newTokenEventName = `newToken`;
    const partialUpdateEventName = "partialUpdate";
    const successEventName = `success`;
    const finishEventName = `finish`;
    const startEventName = `start`;
    const errorEventName = `error`;
    const eventsIterationsMap = /* @__PURE__ */ new Map();
    const startTimeDate = (/* @__PURE__ */ new Date()).getTime();
    const startTimePerf = performance.now();
    function convertDateToPerformance(date) {
      return date.getTime() - startTimeDate + startTimePerf;
    }
    __name(convertDateToPerformance, "convertDateToPerformance");
    const serializer = traceSerializer_cjs.traceSerializer({
      ignored_keys: config_cjs.INSTRUMENTATION_IGNORED_KEYS
    });
    function cleanSpanSources({ spanId }) {
      const parentId = spansMap.get(spanId)?.parent_id;
      if (!parentId) {
        return;
      }
      const spanCount = parentIdsMap.get(parentId);
      if (!spanCount) {
        return;
      }
      if (spanCount > 1) {
        parentIdsMap.set(parentId, spanCount - 1);
      } else if (spanCount === 1) {
        parentIdsMap.delete(parentId);
        if (spansToDeleteMap.has(parentId)) {
          spansMap.delete(parentId);
          spansToDeleteMap.delete(parentId);
        }
      }
    }
    __name(cleanSpanSources, "cleanSpanSources");
    emitter.match((event) => event.path === `${basePath}.run.${finishEventName}`, async (_, meta) => {
      try {
        logger_cjs.instrumentationLogger.debug({
          path: meta.path,
          traceId
        }, "run finish event");
        if (!prompt && instance instanceof base_cjs.BaseAgent) {
          prompt = remeda.findLast(instance.memory.messages, (message) => message.role === message_cjs.Role.USER)?.text;
          if (!prompt) {
            throw new errors_cjs.FrameworkError("The prompt must be defined for the Agent's run", [], {
              context
            });
          }
        }
        tracer_cjs.buildTraceTree({
          prompt,
          history,
          generatedMessage,
          spans: JSON.parse(serializer(Array.from(spansMap.values()))),
          traceId,
          version: version_cjs.Version,
          runErrorSpanKey: `${basePath}.run.${errorEventName}`,
          startTime: startTimePerf,
          endTime: performance.now(),
          source: tracer_cjs.activeTracesMap.get(traceId)
        });
      } catch (e) {
        logger_cjs.instrumentationLogger.warn(e, "Instrumentation error");
      } finally {
        tracer_cjs.activeTracesMap.delete(traceId);
      }
    });
    emitter.match("*.*", (data, meta) => {
      if (meta.path.includes(".run.") && meta.path !== `${basePath}.run.${errorEventName}`) {
        return;
      }
      if (!meta.trace?.runId) {
        throw new errors_cjs.FrameworkError(`Fatal error. Missing runId for event: ${meta.path}`, [], {
          context
        });
      }
      if (meta.groupId && !meta.trace.parentRunId && !groupIterations.includes(meta.groupId)) {
        spansMap.set(meta.groupId, createSpan_cjs.createSpan({
          id: meta.groupId,
          name: meta.groupId,
          target: "groupId",
          startedAt: convertDateToPerformance(meta.createdAt)
        }));
        groupIterations.push(meta.groupId);
      }
      const { spanId, parentSpanId } = idNameManager.getIds({
        path: meta.path,
        id: meta.id,
        runId: meta.trace.runId,
        parentRunId: meta.trace.parentRunId,
        groupId: meta.groupId
      });
      const serializedData = getSerializedObjectSafe_cjs.getSerializedObjectSafe(data);
      if (meta.name === partialUpdateEventName && remeda.isEmpty(serializedData)) {
        return;
      }
      const span = createSpan_cjs.createSpan({
        id: spanId,
        name: meta.name,
        target: meta.path,
        ...parentSpanId && {
          parent: {
            id: parentSpanId
          }
        },
        ctx: getSerializedObjectSafe_cjs.getSerializedObjectSafe(meta.context),
        data: serializedData,
        error: getErrorSafe_cjs.getErrorSafe(data),
        startedAt: convertDateToPerformance(meta.createdAt)
      });
      const lastIteration = groupIterations[groupIterations.length - 1];
      const lastIterationOnNewTokenSpanId = eventsIterationsMap.get(lastIteration)?.get(meta.name);
      if (lastIterationOnNewTokenSpanId && meta.name === newTokenEventName) {
        cleanSpanSources({
          spanId: lastIterationOnNewTokenSpanId
        });
        spansMap.delete(lastIterationOnNewTokenSpanId);
      }
      const lastIterationEventSpanId = eventsIterationsMap.get(lastIteration)?.get(meta.name);
      if (lastIterationEventSpanId && partialUpdateEventName === meta.name && spansMap.has(lastIterationEventSpanId)) {
        const { attributes, context: context2 } = spansMap.get(lastIterationEventSpanId);
        if (remeda.isDeepEqual(serializedData, attributes.data)) {
          if (parentIdsMap.has(context2.span_id)) {
            spansToDeleteMap.set(lastIterationEventSpanId, void 0);
          } else {
            cleanSpanSources({
              spanId: lastIterationEventSpanId
            });
            spansMap.delete(lastIterationEventSpanId);
          }
        }
      }
      spansMap.set(span.context.span_id, span);
      if (span.parent_id) {
        parentIdsMap.set(span.parent_id, (parentIdsMap.get(span.parent_id) || 0) + 1);
      }
      if (groupIterations.length > 0) {
        if (eventsIterationsMap.has(lastIteration)) {
          eventsIterationsMap.get(lastIteration).set(meta.name, span.context.span_id);
        } else {
          eventsIterationsMap.set(lastIteration, (/* @__PURE__ */ new Map()).set(meta.name, span.context.span_id));
        }
      }
    });
    emitter.match((event) => event.name === successEventName && event.creator instanceof base_cjs.BaseAgent, (data) => {
      const { data: dataObject, memory } = data;
      generatedMessage = {
        role: dataObject.role,
        text: dataObject.text
      };
      history = memory.messages.map((msg) => ({
        text: msg.text,
        role: msg.role
      }));
    });
    emitter.match((event) => utils_cjs$1.assertLLMWithMessagesToPromptFn(event.creator) && event.name === startEventName, ({ input }, meta) => {
      if (utils_cjs$1.assertLLMWithMessagesToPromptFn(meta.creator) && meta.trace) {
        const rawPrompt = meta.creator.messagesToPrompt(input);
        const path = `${meta.path}.custom`;
        const { spanId, parentSpanId } = idNameManager.getIds({
          path,
          id: meta.id,
          runId: meta.trace.runId,
          parentRunId: meta.trace.parentRunId,
          groupId: meta.groupId
        });
        spansMap.set(spanId, createSpan_cjs.createSpan({
          id: spanId,
          name: `${meta.name}Custom`,
          target: path,
          startedAt: convertDateToPerformance(meta.createdAt),
          ...parentSpanId && {
            parent: {
              id: parentSpanId
            }
          },
          data: {
            rawPrompt,
            creator: meta.creator.createSnapshot()
          }
        }));
      }
    });
  };
}
__name(createTelemetryMiddleware, "createTelemetryMiddleware");

exports.createTelemetryMiddleware = createTelemetryMiddleware;
