// bee-agent-framework/dist/tools/search/duckDuckGoSearch.js
import { search, SafeSearchType } from 'duck-duck-scrape';
export { SafeSearchType as DuckDuckGoSearchToolSearchType } from 'duck-duck-scrape';
import { stripHtml } from 'string-strip-html';
import pThrottle from 'p-throttle';
import { SearchToolOutput } from './base.js';
import { Tool } from '../base.js';
import { HeaderGenerator } from 'header-generator';
import { z } from 'zod';
import { Cache } from '../../cache/decoratorCache.js';
import { paginate } from '../../internals/helpers/paginate.js';
import { Emitter } from '../../emitter/emitter.js';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
function _ts_decorate(decorators, target, key, desc) {
  var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
  if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
  else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  return c > 3 && r && Object.defineProperty(target, key, r), r;
}
__name(_ts_decorate, "_ts_decorate");
function _ts_metadata(k, v) {
  if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
}
__name(_ts_metadata, "_ts_metadata");
class DuckDuckGoSearchToolOutput extends SearchToolOutput {
  static {
    __name(this, "DuckDuckGoSearchToolOutput");
  }
  results;
  constructor(results) {
    super(results), this.results = results;
  }
  static {
    this.register();
  }
  createSnapshot() {
    return {
      results: this.results
    };
  }
  loadSnapshot(snapshot) {
    Object.assign(this, snapshot);
  }
}
class DuckDuckGoSearchTool extends Tool {
  static {
    __name(this, "DuckDuckGoSearchTool");
  }
  name = "DuckDuckGo";
  description = "Search for online trends, news, current events, real-time information, or research topics.";
  emitter = Emitter.root.child({
    namespace: [
      "tool",
      "search",
      "ddg"
    ],
    creator: this
  });
  client;
  inputSchema() {
    return z.object({
      query: z.string({
        description: `Search query`
      }).min(1).max(128)
    });
  }
  constructor(options = {}) {
    super({
      ...options,
      maxResults: options?.maxResults ?? 15
    });
    this.client = this._createClient();
  }
  static {
    this.register();
  }
  _createClient() {
    const { throttle } = this.options;
    return throttle === false ? search : pThrottle({
      ...throttle,
      limit: throttle?.limit ?? 1,
      interval: throttle?.interval ?? 3e3
    })(search);
  }
  async _run({ query: input }, options, run) {
    const headers = new HeaderGenerator().getHeaders();
    const results = await paginate({
      size: this.options.maxResults,
      handler: /* @__PURE__ */ __name(async ({ offset }) => {
        const { results: data, noResults: done } = await this.client(input, {
          safeSearch: SafeSearchType.MODERATE,
          ...this.options.search,
          ...options?.search,
          offset
        }, {
          headers,
          user_agent: headers["user-agent"],
          ...this.options?.httpClientOptions,
          ...options?.httpClientOptions,
          signal: run.signal,
          uri_modifier: /* @__PURE__ */ __name((rawUrl) => {
            const url = new URL(rawUrl);
            url.searchParams.delete("ss_mkt");
            return url.toString();
          }, "uri_modifier")
        });
        return {
          data,
          done
        };
      }, "handler")
    });
    return new DuckDuckGoSearchToolOutput(results.map((result) => ({
      title: stripHtml(result.title).result,
      description: stripHtml(result.description).result,
      url: result.url
    })));
  }
  loadSnapshot(snapshot) {
    super.loadSnapshot(snapshot);
    Object.assign(this, {
      client: this._createClient()
    });
  }
}
_ts_decorate([
  Cache(),
  _ts_metadata("design:type", Function),
  _ts_metadata("design:paramtypes", []),
  _ts_metadata("design:returntype", void 0)
], DuckDuckGoSearchTool.prototype, "inputSchema", null);

export { DuckDuckGoSearchTool, DuckDuckGoSearchToolOutput };
