// bee-agent-framework/dist/template.cjs
'use strict';

var errors_cjs = require('./errors.cjs');
var R = require('remeda');
var Mustache = require('mustache');
var serializable_cjs = require('./internals/serializable.cjs');
var schema_cjs = require('./internals/helpers/schema.cjs');
var utils_cjs = require('./serializer/utils.cjs');
var object_cjs = require('./internals/helpers/object.cjs');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var R__namespace = /*#__PURE__*/_interopNamespace(R);
var Mustache__default = /*#__PURE__*/_interopDefault(Mustache);

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class PromptTemplateError extends errors_cjs.FrameworkError {
  static {
    __name(this, "PromptTemplateError");
  }
  template;
  constructor(message, template, context) {
    super(message, [], {
      context
    });
    this.template = template;
  }
}
class ValidationPromptTemplateError extends PromptTemplateError {
  static {
    __name(this, "ValidationPromptTemplateError");
  }
}
class PromptTemplate extends serializable_cjs.Serializable {
  static {
    __name(this, "PromptTemplate");
  }
  config;
  static functions = {
    trim: /* @__PURE__ */ __name(() => (text, render) => {
      return render(text).replace(/(,\s*$)/g, "");
    }, "trim")
  };
  constructor(config) {
    super();
    this.config = {
      ...config,
      defaults: config.defaults ?? {},
      schema: schema_cjs.toJsonSchema(config.schema),
      escape: Boolean(config.escape),
      customTags: config.customTags ?? [
        "{{",
        "}}"
      ],
      functions: {
        ...PromptTemplate.functions,
        ...config.functions
      }
    };
  }
  static {
    this.register();
  }
  validateInput(input) {
    const validator = schema_cjs.createSchemaValidator(this.config.schema, {
      coerceTypes: false
    });
    const success = validator(input);
    if (!success) {
      throw new ValidationPromptTemplateError(`Template cannot be rendered because input does not adhere to the template schema.`, this, {
        errors: validator.errors
      });
    }
  }
  fork(customizer) {
    const config = utils_cjs.shallowCopy(this.config);
    const newConfig = customizer?.(config) ?? config;
    return new PromptTemplate(newConfig);
  }
  render(input) {
    const updatedInput = {
      ...input
    };
    Object.assign(updatedInput, R.pickBy(this.config.defaults, (_, k) => object_cjs.getProp(updatedInput, [
      k
    ]) === void 0));
    this.validateInput(updatedInput);
    const view = {
      ...this.config.functions,
      ...updatedInput
    };
    const output = Mustache__default.default.render(this.config.template, view, {}, {
      tags: this.config.customTags,
      ...!this.config.escape && {
        escape: R__namespace.identity()
      }
    });
    if (output.match(/\[object (.+)]/)) {
      throw new ValidationPromptTemplateError("Rendered template contains incorrectly serialized value!", this, {
        output
      });
    }
    return output;
  }
  loadSnapshot(data) {
    this.config = data.config;
  }
  createSnapshot() {
    return {
      config: this.config
    };
  }
}

exports.PromptTemplate = PromptTemplate;
exports.PromptTemplateError = PromptTemplateError;
exports.ValidationPromptTemplateError = ValidationPromptTemplateError;
