// bee-agent-framework/dist/llms/drivers/json.js
import { parseBrokenJson } from '../../internals/helpers/schema.js';
import { PromptTemplate } from '../../template.js';
import { BaseDriver } from './base.js';
import { z } from 'zod';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class JsonDriver extends BaseDriver {
  static {
    __name(this, "JsonDriver");
  }
  template = new PromptTemplate({
    schema: z.object({
      schema: z.string()
    }),
    template: `You are a helpful assistant that generates only valid JSON adhering to the following JSON Schema.

\`\`\`
{{schema}}
\`\`\`

IMPORTANT: Every message must be a parsable JSON string without additional output.
`
  });
  static {
    this.register();
  }
  static fromTemplate(template, ...parameters) {
    const driver = new JsonDriver(...parameters);
    driver.template = template;
    return driver;
  }
  parseResponse(textResponse) {
    return parseBrokenJson(textResponse);
  }
  schemaToString(schema) {
    return JSON.stringify(schema, null, 2);
  }
  guided(schema) {
    return {
      json: schema
    };
  }
}

export { JsonDriver };
